quinn-proto/src/connection/bytes_source.rs

```diff
@@ -0,0 +1,249 @@
+//! Defines types which can be lazily converted into `Bytes` chunks
+
+use bytes::Bytes;
+
+/// A source of one or more buffers which can be converted into `Bytes` buffers on demand.
+///
+/// The purpose of this data type is to defer conversion as long as possible,
+/// so that no heap allocation is required in case no data is writable.
+pub trait BytesSource {
+    // Returns the total amount of bytes to write
+    fn len(&self) -> usize;
+    /// Limits the source to the new size
+    fn limit(&mut self, new_len: usize);
+    /// The amount of bytes consumed from this source
+    fn bytes_consumed(&self) -> usize;
+    /// The amount of chunks consumed from this source
+    fn chunks_consumed(&self) -> usize;
+    /// Whether a chunk was partially consumed from this source
+    fn partial_chunk_consumed(&self) -> bool;
```

Comment:
I feel like the consumed interface is very verbose. Could we maybe simplify it to `consumed(&self) -> (usize, Option<usize>)`?

Reply:
I don't know. It's nice to understand based on the names what is going on. And it would be `consumed(&self) -> (usize, usize, bool)` - since the API can consume e.g. 1.5 buffers. Then users are more likely to mess up the order of arguments.

Reply:
I'm also fine with making it a named struct.

Reply:
Ok. Should this then also be used for the public `quinn-proto` API? any recommendations for name? `BytesWritten` sounds like it contains only bytes. `WriteResult` sounds more like a `Result` type

Reply:
Yeah, I'd be fine with having that as public API, I think.